clientconn_test.go

```diff
@@ -766,3 +767,18 @@ func TestResetConnectBackoff(t *testing.T) {
 		t.Fatal("Failed to call dial within 10s after resetting backoff")
 	}
 }
+
+func TestBackoffCancel(t *testing.T) {
+	defer leakcheck.Check(t)
+	dialStrCh := make(chan string)
+	cc, err := Dial("any", WithInsecure(), WithDialer(func(t string, _ time.Duration) (net.Conn, error) {
+		dialStrCh <- t
+		return nil, fmt.Errorf("test dialer, always error")
+	}))
+	if err != nil {
+		t.Fatalf("Failed to create ClientConn: %v", err)
+	}
+	<-dialStrCh
+	cc.Close()
+	// Should not leak. May need -count 5000 to exercise.
```

Comment:
What was the error rate before fixing the bug?  I was expecting close to 100%.

Reply:
Not super reliable - maybe 1 in 1000. It relies on the race between `addrConn` past the `if ac.state == SHUTDOWN` before the state gets changed in `ac.tearDown`.

Reply:
Ah right -- otherwise it would be 100%.